limitador/src/storage/redis/counters_cache.rs

```diff
@@ -5,25 +5,84 @@ use crate::storage::redis::{
     DEFAULT_TTL_RATIO_CACHED_COUNTERS,
 };
 use moka::sync::Cache;
-use std::sync::Arc;
+use std::collections::HashMap;
+use std::sync::atomic::{AtomicI64, Ordering};
+use std::sync::{Arc, Mutex};
 use std::time::{Duration, SystemTime};
+use tokio::select;
+use tokio::sync::Notify;
+use tokio::time::interval;
 
 pub struct CachedCounterValue {
     value: AtomicExpiringValue,
+    initial_value: AtomicI64,
     expiry: AtomicExpiryTime,
 }
 
+pub struct Batcher {
+    updates: Mutex<HashMap<Counter, Arc<CachedCounterValue>>>,
+    notifier: Notify,
+    interval: Duration,
+}
+
+impl Batcher {
+    fn new(period: Duration) -> Self {
+        Self {
+            updates: Mutex::new(Default::default()),
+            notifier: Default::default(),
+            interval: period,
+        }
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.updates.lock().unwrap().is_empty()
+    }
+
+    pub async fn consume(&self, min: usize) -> HashMap<Counter, Arc<CachedCounterValue>> {
+        let mut interval = interval(self.interval);
+        let mut ready = self.updates.lock().unwrap().len() >= min;
+        loop {
+            if ready {
+                return self.consume_all();
```

Comment:
Here, we'd create a ordered list and return the top `min` entries... So this would also break all the pending updates in smaller batches, it'd probably also need to remove all `0` `CachedCounterValue::pending_writes`...